### PR TITLE
[SPARK-21681][ML] fix bug of MLOR do not work correctly when featureStd contains zero	

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

fix bug of MLOR do not work correctly when featureStd contains zero	

We can reproduce the bug through such dataset (features including zero variance), will generate wrong result (all coefficients becomes 0)
```
    val multinomialDatasetWithZeroVar = {
      val nPoints = 100
      val coefficients = Array(
        -0.57997, 0.912083, -0.371077,
        -0.16624, -0.84355, -0.048509)

      val xMean = Array(5.843, 3.0)
      val xVariance = Array(0.6856, 0.0)  // including zero variance

      val testData = generateMultinomialLogisticInput(
        coefficients, xMean, xVariance, addIntercept = true, nPoints, seed)

      val df = sc.parallelize(testData, 4).toDF().withColumn("weight", lit(1.0))
      df.cache()
      df
    }
```
## How was this patch tested?

testcase added.

### CODE REVIEW

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 1]
```diff
@@ -1392,6 +1415,61 @@ class LogisticRegressionSuite
     assert(model2.interceptVector.toArray.sum ~== 0.0 absTol eps)
   }
 
+  test("test SPARK-21681") {
```
Comment (MrBago) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: I would include a description of the test in addition to the ticket #.

**Review by MrBago — APPROVED** [review]
The only thing I would change is the name of the new test you added. I would add "multinomial logistic regression with zero var" or something similar to the test name.

**Comment by MrBago** [comment]
@jkbradley please take a look when you get a chance.

**Comment by jkbradley** [comment]
LGTM except for making the test's title more descriptive.  Thanks!

**Comment by MLnick** [comment]
cc @sethah

**Comment by MLnick** [comment]
Fix seems ok - but just wondering why the test for zero std dev in `LogisticAggregatorSuite` didn't pick this up? cc @sethah 

https://github.com/apache/spark/blob/cf29828d720611f7633a8114917bac901f76dece/mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala#L232

**Comment by WeichenXu123** [comment]
@MLnick That's because, this bug will be triggered only when we standardize feature first then do training...

**Comment by MLnick** [comment]
The model is always trained in the scaled space regardless of the `standardization` param.

e.g. in the test case doing:
```scala
    val model = mlr.fit(multinomialDatasetWithZeroVar)

    println(model.interceptVector)
    println(model.coefficientMatrix)

    val model2 = mlr.setStandardization(false).fit(multinomialDatasetWithZeroVar)

    println(model2.interceptVector)
    println(model2.coefficientMatrix)
```

Gives you
```
[0.2659042827402973,0.5361544814545004,-0.8020587641947977]
0.18818263258382864    0.0  
-0.024143986869901088  0.0  
-0.16403864571392754   0.0  
[0.2659042827402915,0.5361544814544981,-0.8020587641947895]
0.18818263258382964    0.0  
-0.024143986869900692  0.0  
-0.16403864571392893   0.0  
```

**Comment by WeichenXu123** [comment]
@MLnick Yes it is always trained in scaled space. But the testcase you mentioned do not take the "scale" step, so do not trigger the bug...

**Comment by WeichenXu123** [comment]
@MLnick I debug the testcase your mentioned.
The reason is, the zero var cause the computation generate `Infinite` and `NaN` so the result is unexpectable, in this case, it happened to generate 0.0 at gradient(0), so it escape the testcase.

**Comment by jkbradley** [comment]
How about updating the LogisticAggregatorSuite so it catches this error:
```
diff --git a/mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala b/mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala
index 2b29c67..a2924ad 100644
--- a/mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala
+++ b/mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala
@@ -239,7 +239,14 @@ class LogisticAggregatorSuite extends SparkFunSuite with MLlibTestSparkContext {
       Vectors.dense(coefArray ++ interceptArray), fitIntercept = true, isMultinomial = true)
     instances.foreach(aggConstantFeature.add)
     // constant features should not affect gradient
-    assert(aggConstantFeature.gradient(0) === 0.0)
+    def validateGradient(grad: Vector): Unit = {
+      assert(grad(0) === 0.0)
+      grad.toArray.foreach { gradientValue =>
+        assert(!gradientValue.isNaN &&
+          gradientValue > Double.NegativeInfinity && gradientValue < Double.PositiveInfinity)
+      }
+    }
+    validateGradient(aggConstantFeature.gradient)
 
     val binaryCoefArray = Array(1.0, 2.0)
     val intercept = 1.0
@@ -248,6 +255,6 @@ class LogisticAggregatorSuite extends SparkFunSuite with MLlibTestSparkContext {
       isMultinomial = false)
     instances.foreach(aggConstantFeatureBinary.add)
     // constant features should not affect gradient
-    assert(aggConstantFeatureBinary.gradient(0) === 0.0)
+    validateGradient(aggConstantFeatureBinary.gradient)
   }
 }
```

**Comment by WeichenXu123** [comment]
@jkkbradley OK. So I can remove the test I added ?

File: mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala [hunk 2]
```diff
@@ -238,8 +238,17 @@ class LogisticAggregatorSuite extends SparkFunSuite with MLlibTestSparkContext {
     val aggConstantFeature = getNewAggregator(instancesConstantFeature,
       Vectors.dense(coefArray ++ interceptArray), fitIntercept = true, isMultinomial = true)
     instances.foreach(aggConstantFeature.add)
+
     // constant features should not affect gradient
-    assert(aggConstantFeature.gradient(0) === 0.0)
+    def validateGradient(grad: Vector): Unit = {
+      assert(grad(0) === 0.0)
+      grad.toArray.foreach { gradientValue =>
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/optim/aggregator/LogisticAggregatorSuite.scala]: The problem with this test was that it checked that part of the gradient was zero, but didn't check that the rest of the gradient was correct. Here, you're checking that the rest of the gradient isn't nan or infinite, but not that it's actually correct. A more appropriate test, IMO, is to also run an aggregator over the same instances with the constant feature filtered out, then check that the portion of the gradients they share are the same. e.g.

````scala
    val aggConstantFeature = getNewAggregator(instancesConstantFeature,
      Vectors.dense(coefArray ++ interceptArray), fitIntercept = true, isMultinomial = true)
    val filteredInstances = instancesConstantFeature.map { case Instance(l, w, f) =>
      Instance(l, w, Vectors.dense(f.toArray.tail))
    }
    val aggMultinomial = getNewAggregator(filteredInstances,
      Vectors.dense(coefArray.slice(3, 6) ++ interceptArray), fitIntercept = true,
      isMultinomial = true)
    filteredInstances.foreach(aggMultinomial.add)
    instancesConstantFeature.foreach(aggConstantFeature.add)

    // constant features should not affect gradient
    assert(aggConstantFeature.gradient.toArray.take(numClasses) === Array.fill(numClasses)(0.0))
    assert(aggMultinomial.gradient.toArray === aggConstantFeature.gradient.toArray.slice(3, 9))
````

Just to note, this code is just for an example, not meant to be copy and pasted.

File: mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala [hunk 3]
```diff
@@ -1392,6 +1415,61 @@ class LogisticRegressionSuite
...
+     $`2`
+     3 x 1 sparse Matrix of class "dgCMatrix"
+                     s0
+             -0.8019294
+     data.V3 -0.1640607
+     data.V4  .
+    */
+
+    val coefficientsR = new DenseMatrix(3, 2, Array(
+      0.1881871, -0.0,
```
Comment (sethah) [inline mllib/src/test/scala/org/apache/spark/ml/classification/LogisticRegressionSuite.scala]: Why `-0.0`?

**Comment by sethah** [comment]
Thanks for catching this @WeichenXu123! I just added a note about the intent of test.

**Comment by jkbradley** [comment]
Thanks @WeichenXu123 and @sethah !

@WeichenXu123 I'd leave the MLOR test since it's cheap and has a clear purpose, even if it overlaps a little.

LGTM
@sethah any more comments before I merge it?

**Comment by jkbradley** [comment]
Merging with master and branch-2.2
Thanks @WeichenXu123 @MLnick @sethah !

**Comment by jkbradley** [comment]
@WeichenXu123 would you mind sending a backport PR for 2.2?

**Comment by WeichenXu123** [comment]
@jkbradley OK. (Can this directly merged to 2.2 ?)

**Pull request closed without merging**